CSharpier crashes when run mutliple times simultaneously

Hi!

In my team's project, we use the `CSharpier.MsBuild` project for build-time formatting. We include it in all our projects by way of `Directory.Build.props` file in the repo root. 

As of v0.19, CSharpier fails when attempting to rebuild the entire solution (about 7 projects), because multiple instances of the tool are attempting to write to the cache file, and cannot get exclusive access. Here's an example of the failure from an MSBuild log:

```
Target CSharpierFormat:
4>  Unhandled exception: System.IO.IOException: The process cannot access the file 'C:\Users\userdir\AppData\Local\CSharpier\.formattingCache' because it is being used by another process.
6>Target CSharpierFormat:
6>  Total time:                                                                          432ms
6>  Total files:                                                                          15
4>     at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
4>     at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
4>     at System.IO.Strategies.OSFileStreamStrategy..ctor(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
4>     at System.IO.Strategies.FileStreamHelpers.ChooseStrategyCore(String path, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
4>     at System.IO.Strategies.FileStreamHelpers.ChooseStrategy(FileStream fileStream, String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, Int64 preallocationSize)
4>     at System.IO.File.AsyncStreamWriter(String path, Encoding encoding, Boolean append)
4>     at System.IO.File.WriteAllTextAsync(String path, String contents, Encoding encoding, CancellationToken cancellationToken)
4>     at System.IO.File.WriteAllTextAsync(String path, String contents, CancellationToken cancellationToken)
4>     at System.IO.Abstractions.FileWrapper.WriteAllTextAsync(String path, String contents, CancellationToken cancellationToken)
4>     at CSharpier.Cli.FormattingCacheFactory.FormattingCache.ResolveAsync(CancellationToken cancellationToken) in /home/runner/work/csharpier/csharpier/Src/CSharpier.Cli/FormattingCache.cs:line 113
4>     at CSharpier.Cli.CommandLineFormatter.FormatPhysicalFiles(CommandLineFormatterResult commandLineFormatterResult, CommandLineOptions commandLineOptions, IFileSystem fileSystem, IConsole console, ILogger logger, CancellationToken cancellationToken) in /home/runner/work/csharpier/csharpier/Src/CSharpier.Cli/CommandLineFormatter.cs:line 220
4>     at CSharpier.Cli.CommandLineFormatter.Format(CommandLineOptions commandLineOptions, IFileSystem fileSystem, IConsole console, ILogger logger, CancellationToken cancellationToken) in /home/runner/work/csharpier/csharpier/Src/CSharpier.Cli/CommandLineFormatter.cs:line 63
4>     at CSharpier.Cli.Program.Run(String[] directoryOrFile, Boolean check, Boolean fast, Boolean skipWrite, Boolean writeStdout, Boolean pipeMultipleFiles, Boolean noCache, CancellationToken cancellationToken) in /home/runner/work/csharpier/csharpier/Src/CSharpier.Cli/Program.cs:line 81
4>     at System.CommandLine.Invocation.CommandHandler.GetExitCodeAsync(Object value, InvocationContext context)
4>     at System.CommandLine.Invocation.ModelBindingCommandHandler.InvokeAsync(InvocationContext context)
4>     at System.CommandLine.Invocation.InvocationPipeline.<>c__DisplayClass4_0.<<BuildInvocationChain>b__0>d.MoveNext()
4>  --- End of stack trace from previous location ---
4>     at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass23_0.<<UseParseErrorReporting>b__0>d.MoveNext()
4>  --- End of stack trace from previous location ---
4>     at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass16_0.<<UseHelp>b__0>d.MoveNext()
4>  --- End of stack trace from previous location ---
4>     at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass27_0.<<UseVersionOption>b__1>d.MoveNext()
4>  --- End of stack trace from previous location ---
4>     at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass25_0.<<UseTypoCorrections>b__0>d.MoveNext()
4>  --- End of stack trace from previous location ---
4>     at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<UseSuggestDirective>b__24_0>d.MoveNext()
4>  --- End of stack trace from previous location ---
4>     at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass22_0.<<UseParseDirective>b__0>d.MoveNext()
4>  --- End of stack trace from previous location ---
4>     at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass11_0.<<UseDebugDirective>b__0>d.MoveNext()
4>  --- End of stack trace from previous location ---
4>     at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c.<<RegisterWithDotnetSuggest>b__10_0>d.MoveNext()
4>  --- End of stack trace from previous location ---
4>     at System.CommandLine.Builder.CommandLineBuilderExtensions.<>c__DisplayClass14_0.<<UseExceptionHandler>b__0>d.MoveNext()
4>  C:\Users\userdir\.nuget\packages\csharpier.msbuild\0.19.0\build\CSharpier.MsBuild.targets(11,9): error MSB3073: The command "dotnet C:\Users\userdir\.nuget\packages\csharpier.msbuild\0.19.0\build\../tools/csharpier/dotnet-csharpier.dll  C:\Users\userdir\source_code\project\src\project" exited with code 1.
4
```

Because it's a contention issue, it _can_ manifest with as few as two projects in a solution, but it's much more likely once you start trying to build 5 or more simultaneously.

--- a/Src/CSharpier.Cli.Tests/CliTests.cs
+++ b/Src/CSharpier.Cli.Tests/CliTests.cs
@@ -10,6 +10,10 @@ using NUnit.Framework;
 
 namespace CSharpier.Cli.Tests;
 
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection.Metadata;
+
 // these tests are kind of nice as c# because they run in the same place.
 // except the one test that has issues with console input redirection
 // they used to be powershell, but doing the multiple file thing didn't work
@@ -313,6 +317,39 @@ public class CliTests
 
         var result = await this.ReadAllTextAsync("Unformatted.cs");
         result.Should().Contain("\n\t// break\n");
+    }
+
+    [Test]
+    public async Task Should_Handle_Concurrent_Processes()
+    {
+        var unformattedContent = "public class ClassName {     }\n";
+        var totalFolders = 10;
+        var filesPerFolder = 100;
+        var folders = new List<string>();
+        for (var x = 0; x < totalFolders; x++)
+        {
+            folders.Add("Folder" + x);
+        }
+
+        async Task WriteFiles(string folder)
+        {
+            for (var y = 0; y < filesPerFolder; y++)
+            {
+                await this.WriteFileAsync($"{folder}/File{y}.cs", unformattedContent);
+            }
+        }
+
+        var tasks = folders.Select(WriteFiles).ToArray();
+        Task.WaitAll(tasks);
+
+        async Task FormatFolder(string folder)
+        {
+            var result = await new CsharpierProcess().WithArguments(folder).ExecuteAsync();
+            result.ErrorOutput.Should().BeEmpty();
+        }
+
+        var formatTasks = folders.Select(FormatFolder).ToArray();
+        Task.WaitAll(formatTasks);
     }
 
     private static bool CannotRunTestWithRedirectedInput()

--- a/Src/CSharpier.Cli/FormattingCache.cs
+++ b/Src/CSharpier.Cli/FormattingCache.cs
@@ -110,17 +110,35 @@ internal static class FormattingCacheFactory
         {
             this.fileSystem.FileInfo.FromFileName(this.cacheFile).EnsureDirectoryExists();
 
-            await this.fileSystem.File.WriteAllTextAsync(
-                this.cacheFile,
-                JsonSerializer.Serialize(
-                    this.cacheDictionary
+            async Task WriteFile()
+            {
+                await this.fileSystem.File.WriteAllTextAsync(
+                    this.cacheFile,
+                    JsonSerializer.Serialize(
+                        this.cacheDictionary
 #if DEBUG
-                    ,
-                    new JsonSerializerOptions { WriteIndented = true }
+                        ,
+                        new JsonSerializerOptions { WriteIndented = true }
 #endif
-                ),
-                cancellationToken
-            );
+                    ),
+                    cancellationToken
+                );
+            }
+
+            var wait = 1;
+            for (var x = 0; x < 10; x++)
+            {
+                try
+                {
+                    await WriteFile();
+                    return;
+                }
+                catch (Exception)
+                {
+                    await Task.Delay(TimeSpan.FromMilliseconds(wait), cancellationToken);
+                    wait *= 2;
+                }
+            }
         }
     }
 

--- a/Src/CSharpier.Cli/FormattingCache.cs
+++ b/Src/CSharpier.Cli/FormattingCache.cs
@@ -125,8 +125,8 @@ internal static class FormattingCacheFactory
                 );
             }
 
-            var wait = 1;
-            for (var x = 0; x < 10; x++)
+            // in my testing we don't normally have to wait more than a couple MS, but just in case
+            for (var x = 0; x < 20; x++)
             {
                 try
                 {
@@ -135,8 +135,7 @@ internal static class FormattingCacheFactory
                 }
                 catch (Exception)
                 {
-                    await Task.Delay(TimeSpan.FromMilliseconds(wait), cancellationToken);
-                    wait *= 2;
+                    await Task.Delay(TimeSpan.FromMilliseconds(1), cancellationToken);
                 }
             }
         }

--- a/Src/CSharpier/DebugLogger.cs
+++ b/Src/CSharpier/DebugLogger.cs
@@ -17,7 +17,6 @@ public class DebugLogger
             }
             catch (Exception)
             {
-                File.AppendAllText(@"C:\projects\csharpier\debug.txt", message + "\n");
                 // we don't care if this fails
             }
         }

